networkx/algorithms/distance_measures.py

```diff
@@ -14,7 +14,7 @@
 ]
 
 
-def _extrema_bounding(G, compute="diameter"):
+def _extrema_bounding(G, compute="diameter", weight="weight"):
```

Comment:
Yes -- it should default to `weight=None`.  

The `weighted.py` module does use `weight="weight"`, but we shouldn't use the `weighted.py` functions unless a `weight` argument is provided. That's what `shortest_path-lengths` does for us (which is in the `generic.py` module). It chooses whether to call the weighted functions, or stick with the unweighted functions with the default choice being unweighted (unless a weight value is provided).

Reply:
> That's what shortest_path_lengths does for us (which is in the generic.py module). It chooses whether to call the weighted functions, or stick with the unweighted functions with the default choice being unweighted (unless a weight value is provided).

I'm not sure I agree @dschult @rossbar . We are trying to distinguish three cases:
1. Weights are via a provided weight function
2. Weights are via edge attributes ( weight="weight")
3. Weights are all 1 ("unweighted" case)

When weight="weight", we're not defaulting to the unweighted case, we're using the edge attributes as weights. If we pass weight=None, we are telling `shortest_path_length` to set weights to 1, not use the edge attributes. This is written in the `shortest_path_length` docstring:

> weight : None, string or function, optional (default = None)
        **If None, every edge has weight/distance/cost 1.**
        If a string, use this edge attribute as the edge weight.
        Any edge attribute not present defaults to 1.
        If this is a function, the weight of an edge is the value
        returned by the function. The function must accept exactly
        three positional arguments: the two endpoints of an edge and
        the dictionary of edge attributes for that edge.
        The function must return a number.

Constructively, three paths forward come to mind:
1. Leave  weight="weight" as the default in all `distance_measures`. Consider refactoring `distance_measures` and `generic` at a later time.
2. In all `_extrema_bounding`, do something like:
```
if not weight:
      dist = nx.shortest_path_length(G=G, source=current, weight="weight")
else:
      dist = nx.shortest_path_length(G=G, source=current, weight=weight)
```
3. refactor `generic` as part of this PR.

My view is refactoring `generic` is out of scope, and #2 seems clunky, but I am happy to do it if you think that best. Do you see a better alternative?

Reply:
We want the unweighted values when the user does not provide the `weight` argument to the function.
I think all the code is just fine except for the default value of weight, which should be None just as it is for `shortest_path_length`.  Then for each use case the right thing happens:

1) Weight=function: The user specifies a weight function, so `weight` is not None and `shortest_path_length` uses the weight provided weight function -- all is good.
2) Weights=edge attribute name.  For example, weight="weight" or weight="distance" or weight="cost" or any other attribute name.  Again, weight is not None, so `shortest_path_length` uses the edge attribute specified to find the weights.
3) Unweighted case:  This is the old/existing case. The user doesn't specify any `weight` and so `weight=None` is passed into `shortest_path_length` which calls the unweighted path length function.

Reply:
@lucasmccabe consider the following.

Let's say a user has the following code:

```python
>>> import networkx as nx
>>> G = nx.Graph([(0, 1, {"weight": 10})])
>>> nx.diameter(G)
```

In all previous versions of networkx, this would give `1`. However, if we were to make `weight="weight"` the default, then this would give `10` instead; i.e. we'd be silently changing the results of existing user code without warning. This is something that we must avoid, otherwise we'd potentially be introducing bugs (and hard-to-track-down ones, at that) all over existing user code!

I certainly understand the logic of wanting a weighted result by default if the graph is weighted. Even without backwards-compatibility constraints, choosing an explicit default for an edge attribute doesn't mesh well with the generality of the library. Consider another example: let's say someone has a graph where the nodes represent towns and the edges represent roads between them (as I gather, this is a pretty common example in geospatial applications). The edges of this graph might have multiple attributes, say one called `"distance"` that represents the physical distance between nodes along a given edge, and another called `"weight"` which represents the number of travelers / hour that travel along that edge. In this case, if a user is calling one of the `distance_measures` functions on the graph, chances are they actually want the result to be weighted by the `"distance"` edge attr, but we'd be giving them a different (and harder to understand for this example) value *by default*. In this case too, I'd argue it's much better to have an unweighted default and force the user to be intentional about which weight they want.

Anyways, just some high-level pontification. The real blocker for me with going with `weight="weight"` as the default is the backward-incompatibility concern!

Reply:
@dschult  @rossbar  Thank you both for being patient and explaining your reasoning here; I now understand what you mean.

I thought you wanted to make `weight=None` do what `weight="weight"` does, and drop `weight="weight"` as an option altogether. My mistake!